src/main/java/org/broadinstitute/hellbender/engine/ReadShard.java

```diff
@@ -0,0 +1,182 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.engine.filters.ReadFilter;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.iterators.ReadFilteringIterator;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * A class to represent a shard of reads data, optionally expanded by a configurable amount of padded data.
+ *
+ * The reads are lazily loaded by default (when accessing the reads via {@link #iterator}. Loading all the
+ * reads in the window at once is possible via {@link #loadAllReads}.
+ *
+ * The reads returned will overlap the expanded padded interval. It's possible to query whether they are within
+ * the main part of the shard via {@link #isContainedWithinShard} and {@link #startsWithinShard}.
+ *
+ * The reads in the window can be filtered via {@link #setReadFilter} (no filtering is performed by default).
+ */
+public final class ReadShard implements Iterable<GATKRead>, Locatable {
+
+    private final SimpleInterval interval;
+    private final SimpleInterval paddedInterval;
+    private final ReadsDataSource readsSource;
+    private ReadFilter readFilter;
+
+    /**
+     * Create a new ReadShard spanning the specified interval, with the specified amount of padding.
+     *
+     * @param interval the genomic span covered by this shard
+     * @param paddedInterval the span covered by this shard, plus any additional padding on each side (must contain the un-padded interval)
+     * @param readsSource source of reads from which to populate this shard
+     */
+    public ReadShard( final SimpleInterval interval, final SimpleInterval paddedInterval, final ReadsDataSource readsSource ) {
+        Utils.nonNull(interval);
+        Utils.nonNull(paddedInterval);
+        Utils.nonNull(readsSource);
+
+        if ( ! paddedInterval.contains(interval) ) {
+            throw new IllegalArgumentException("The padded interval must contain the un-padded interval");
+        }
+
+        this.interval = interval;
+        this.paddedInterval = paddedInterval;
+        this.readsSource = readsSource;
+    }
+
+    /**
+     * Reads in this shard will be filtered using this filter before being returned
+     *
+     * @param filter filter to use (may be null)
+     */
+    public void setReadFilter( final ReadFilter filter ) {
```

Comment:
why is this a setter rather than a constructor parameter? do you want to adjust filtering after the shard is created?